### DISCUSSION TITLE
Logout in custom Authentication Flow SP

### DISCUSSION THREAD

**Discussion by CreepyJoker** [keycloak/keycloak#23968] [Q&A]
Greetings,

So I have created a custom service provider that does the following: `client_id 123` initiates an authentication request, the user is greeted with the UsernamePasswordForm challenge and if the user exists, the user gets an OTP and logs in, gives consent etc. If the user doesn't exists, it's redirected to the KYC solution which then provides back to keycloak the user details, creates the user, sets the user in session and continues with authentication process.

My current goal is that I want to make it so whenever the "client_id" 123 initiates an authentication request and there's already a user authenticated in the browser, to logout the already authenticated user and continue with normal authentication for the new-coming user. 
Some clients do not want to use the logout endpoint before initiating a new request because of a few reasons:
1. They don't want users to "confirm any logout";
2. Sessions may expire and they don't want the users to see the keycloak's "go back to the application" notice as the authentication flow can be part of another application and integrated with a webview;

The problem is that if I simply use the following:
`KeycloakSession session = context.getSession();
session.sessions().removeUserSession(realm, userSession);`
It keeps the same session_code (keycloak's session ID - `context.getAuthenticationSession().getParentSession().getId();)` which is not optimal due to security concerns and other things.

Not being a keycloak expert by any means, I've put down what I've tried so far:

 **as a last execution provider** (last authentication flow step):
1. To force a logout using AuthenticationManager's browserLogout / backchannelLogout functions - same result, keeps the session_code;
4. Replicate the browser logout function bare bones as much as I could without the extra checks for testing purposes - same result;
5. Using `CookieHelper.addCookie` in order to expire the already existing cookies and hopefully get a new session_code when a new auth request is made - works like half-way, as even with a `Browser Redirect/Refresh` still has the same session_code;
6. `session.sessions().removeUserSession(realm, userSession);
session.authenticationSessions().removeRootAuthenticationSession(userSession.getRealm(), context.getAuthenticationSession().getParentSession());` - invalidates the whole session and I get an `error=invalid_code` which should be kind of normal and if I try afterwards to create another rootAuthSession, I get a `Can't add session: task in progress for session` exception.

**as a starting execution provider** (first step in the auth flow):
1. Backchannel / frontend logout that the LogoutEndpoints use;
2. Cookie reset, `session.sessions().removeUserSession(realm, userSession)` & `Browser Redirect/Refresh` 
3. Replication of the LogoutEndpoints & AuthenticationManager's logout logics
4. the exact same 4th point mentioned before
7. Multiple combinations that I can't quite remember which were kind of sketchy

At first glance I thought that expiring cookies, removing the user's existing session and forcing a refresh it will set a new session_code and such, but digging deeper into the source code I've seen that's not quite the case.

Is this thing even achievable using a AuthenticationFlowContext or am I beating stones in hopes that some water will come out?
If not, how can I properly do it? 
My goal was to use it as an extra execution for specific flows, assign a role let's say `auto_logout` to specific clients and check if there should be an auto-logout for a client or not to perform this action, conditionally.

Thanks in advance.

**Comment by danielFesenmeyer** [comment]
We had a similar requirement to force a re-login for certain clients.

Our approach works like that:
- We create a "fake" (but valid) ID token for the requested client `client_2`, based on one of the currently authenticated client sessions (e.g. `client_1`) and on the access token used in the current request (issued for `client_2`)
- We redirect the user to the logout uri of `client_2`, containing the login uri of `client_1` as `post_logout_redirect_uri` and the "fake" ID token as `id_token_hint`

Maybe there are more efficient ways to do this, but it seemed to be the solution which was easiest to implement.
Downside: For this to work, the login uri of `client_1` must be configured as valid _post logout redirect uri_ of `client_2`.

**Comment by Pulkit-m** [comment]
@CreepyJoker Did you ever find the proper way to do this. I am facing similar issue, where I want to force user logout post credential reset. 
I am working with the exact same snippets that you mentioned above.
